pkg/kube/client/ssa/gvk_parser_cache.go

```diff
@@ -0,0 +1,76 @@
+// SPDX-FileCopyrightText: 2024 The Crossplane Authors <https://crossplane.io>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package ssa
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/crossplane-contrib/provider-kubernetes/apis/v1alpha1"
+)
+
+// GVKParserCacheManager maintains GVK parser cache stores for each provider config.
+type GVKParserCacheManager struct {
+	// mu is used to make sure the cacheStore map is concurrency-safe.
+	mu sync.RWMutex
+	// cacheStore holds the *GVKParserCache per provider configuration.
+	// The cacheStore key is the UID of the provider config object.
+	cacheStore map[types.UID]*GVKParserCache
+}
+
+// GVKParserCacheManagerOption lets you configure a *GVKParserCacheManager.
+type GVKParserCacheManagerOption func(cache *GVKParserCacheManager)
+
+// NewGVKParserCacheManager returns a new empty *GVKParserCacheManager.
+func NewGVKParserCacheManager(opts ...GVKParserCacheManagerOption) *GVKParserCacheManager {
+	c := &GVKParserCacheManager{
+		cacheStore: map[types.UID]*GVKParserCache{},
+	}
+	for _, f := range opts {
+		f(c)
+	}
+	return c
+}
+
+// LoadOrNewCacheForProviderConfig returns the *GVKParserCache for the given provider config,
+// initializing an empty cache for the first use.
+// the implementation is concurrency-safe.
+func (cm *GVKParserCacheManager) LoadOrNewCacheForProviderConfig(pc *v1alpha1.ProviderConfig) (*GVKParserCache, error) {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	sc, ok := cm.cacheStore[pc.GetUID()]
```

Comment:
why is the uid enough? Couldn't the content change?

Reply:
cache manager stores the collection of caches per provider config.
The actual cache is the one that is returned, this is done at Connect time and given to the reconciliation scope.
Then the content changes are handled there.

Reply:
```
At connect time you call `LoadOrNewCacheForProviderConfig`, but it will return the same object on every reconcile. If the ProviderConfig is changes, the UID will still be the same and with that the cache is potentially invalid. What do I miss?

Reply:
> At connect time you call LoadOrNewCacheForProviderConfig, but it will return the same object on every reconcile.
yes, we get the same GvkParserCache here for the PC. `LoadOrNewCacheForProviderConfig` just gives you the actual cache to maintain. We can think of `GvkParserCacheManager` like a mere storage organizer of caches, it just points you to the right drawer reserved for PC. Then we pass the GvkParserCache we got to the extractor at Connect time.

The cache itself is maintained after that point by the extractor [here](https://github.com/crossplane-contrib/provider-kubernetes/pull/246/files#diff-a3b611789875d8619ff3bd58cacbf0cea61cd467e8fcdb294936358addb0cd8eR98)

at each extract request:
A v3 discovery is initiated that returns a collection of all served gv paths to Etagged API paths, then according to the output :
- stale entries are invalidated (non-matching Etags, APIs not being served anymore) 
- then we retrieve the schema (if it is not on the cache after invalidations) for the relevant GV of object that is being extracted, and cache it with its ETag.

So, in the case of the ProviderConfig change, the returned discovery information will change, as the discovery client itself is rebuilt at each reconcile. Then, the invalidations will take place, so we operate on up-to-date info.

Reply:
Ack. Seems sounds. This is crucial information that needs to be in some comments